vnext/ReactUWP/Views/KeyboardEventHandler.cpp

```diff
@@ -0,0 +1,402 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "pch.h"
+#include "Views/KeyboardEventHandler.h"
+#include "Utils/Helpers.h"
+#include "Utils/PropertyHandlerUtils.h"
+#include <winrt/Windows.UI.Core.h>
+
+using namespace std::placeholders;
+
+static constexpr auto ALT_KEY = "altKey";
+static constexpr auto CTRL_KEY = "ctrlKey";
+static constexpr auto META_KEY = "metaKey";
+static constexpr auto SHIFT_KEY = "shiftKey";
+static constexpr auto EVENT_PHASE = "handledEventPhase";
+static constexpr auto KEY = "key";
+static constexpr auto TARGET = "target";
+
+using namespace react::uwp;
+
+template<>
+struct json_type_traits<react::uwp::KeyboardEvent>
+{
+  static react::uwp::KeyboardEvent parseJson(const folly::dynamic& json)
+  {
+    react::uwp::KeyboardEvent ev;
+
+    for (auto const& pair : json.items())
+    {
+      const std::string& propertyName = pair.first.getString();
+      const folly::dynamic& propertyValue = pair.second;
+
+      if (propertyName == ALT_KEY)
+        ev.altKey = propertyValue.asBool();
+      else if (propertyName == SHIFT_KEY)
+        ev.shiftKey = propertyValue.asBool();
+      else if (propertyName == CTRL_KEY)
+        ev.ctrlKey = propertyValue.asBool();
+      else if (propertyName == META_KEY)
+        ev.metaKey = propertyValue.asBool();
+      else if (propertyName == KEY)
+        ev.key = propertyValue.asString();
+      else if (propertyName == EVENT_PHASE)
+        ev.handledEventPhase = asEnum<HandledEventPhase>(propertyValue);
+    }
+    return ev;
+  }
+};
+
+namespace react {namespace uwp {
+
+std::vector<KeyboardEvent> KeyboardHelper::FromJS(folly::dynamic const& obj)
+{
+  return json_type_traits<std::vector<KeyboardEvent>>::parseJson(obj);
+}
+
+static folly::dynamic ToEventData(ReactKeyboardEvent event)
+{
+  return folly::dynamic::object
+  (TARGET, event.target)
+    (ALT_KEY, event.altKey)
+    (CTRL_KEY, event.ctrlKey)
+    (KEY, event.key)
+    (META_KEY, event.metaKey)
+    (SHIFT_KEY, event.shiftKey);
+}
+
+KeyboardEventBaseHandler::KeyboardEventBaseHandler(KeyboardEventCallback&& keyDown, KeyboardEventCallback&& keyUp)
+  :m_keyDownCallback(std::move(keyDown)), m_keyUpCallback(std::move(keyUp))
+{
+}
+
+PreviewKeyboardEventHandler::PreviewKeyboardEventHandler(KeyboardEventCallback&& keyDown, KeyboardEventCallback&& keyUp)
+  : KeyboardEventBaseHandler(std::move(keyDown), std::move(keyUp))
+{
+}
+
+void PreviewKeyboardEventHandler::hook(XamlView xamlView)
+{
+  auto uiElement = xamlView.as<winrt::UIElement>();
```

Comment:
>xamlView.as<winrt::UIElement>(); [](start = 19, length = 32)

Are we guaranteed this will be UIElement if this is called?

Reply:
Looks like you are checking this in the other hook implementation, actually!

---
In reply to: [295535258](https://github.com/microsoft/react-native-windows/pull/2623#discussion_r295535258) [](ancestors = 295535258)

Reply:
As I know, yes. as other than try_as is on purpose, and I expect app to crash if we hook to non UIElement.

Reply:
I will change it to as.

---
In reply to: [295537264](https://github.com/microsoft/react-native-windows/pull/2623#discussion_r295537264) [](ancestors = 295537264,295535258)